File: classic2D/classic_2D.m
function classic_2D

N = 100000;       %总粒子数
Nstep = 200;      %总步数

X = round(linspace(-Nstep,Nstep,2*Nstep+1));  %粒子所有可能到达的位置(横轴)
Y = round(linspace(-Nstep,Nstep,2*Nstep+1));  %粒子所有可能到达的位置(纵轴)
x = zeros(1,N);  %每个粒子的横坐标
y = zeros(1,N);  %每个粒子的纵坐标

%> 固定步数Nstep, 观察最后的粒子分布
% tic;
% r = rand(Nstep,N);
% Z = zeros(2*Nstep+1,2*Nstep+1);  %统计最终坐标上的粒子数
% for i = 1:N
%     for j = 1:Nstep
%         if(r(j,i) <= 0.25)
%             x(i) = x(i)-1;
%         elseif(r(j,i) > 0.25 && r(j,i) <= 0.5)
%             x(i) = x(i)+1;
%         elseif(r(j,i) > 0.5 && r(j,i) <= 0.75 )
%             y(i) = y(i)-1;
%         else
%             y(i) = y(i)+1;
%         end
%     end
%     index_x = find(X == x(i));
%     index_y = find(Y == y(i));
%     Z(index_x,index_y) = Z(index_x,index_y) + 1;
% end
% p = Z/N;
% toc
% mesh(X,Y,p)


%> 观察粒子分布随时间(步数)的演化
tic;
r = rand(N,Nstep);
Z = zeros(Nstep,2*Nstep+1,2*Nstep+1);  %统计最终坐标上的粒子数
for i = 1:Nstep
    for j = 1:N
        if(r(j,i) <= 0.25)
            x(j) = x(j)-1;
        elseif(r(j,i) > 0.25 && r(j,i) <= 0.5)
            x(j) = x(j)+1;
        elseif(r(j,i) > 0.5 && r(j,i) <= 0.75)
            y(j) = y(j)-1;
        else
            y(j) = y(j)+1;
        end
        index_x = find(X==x(j));
        index_y = find(Y==y(j));
        Z(i,index_x,index_y) = Z(i,index_x,index_y) + 1;
    end
end
p = Z/N;
toc


%> 单个粒子随机游走
tic;
x0 = 0;
y0 = 0;
r0 = rand(1,Nstep);
site = zeros(Nstep,2*Nstep+1,2*Nstep+1);
for i = 1:Nstep
    if(r0(i)<0.25)
        x0 = x0-1;
    elseif(r0(i)>0.25 && r0(i)<0.5)
        x0 = x0+1;
    elseif(r0(i)>0.5 && r0(i)<0.75)
        y0 = y0-1;
    else
        y0 = y0+1;
    end
    index_x = find(X==x0);
    index_y = find(Y==y0);
    site(i,index_x,index_y) = 1;
end
toc


save("C:\Users\JAH\Desktop\组\random walk\classic2D\classic2D.mat","p","X","Y","Nstep","site");
fprintf("数据已导出");

File: classic2D/classic_2D.py
from os import X_OK
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
import scipy.io as sio
from matplotlib import animation
from matplotlib.animation import FuncAnimation
from matplotlib import colors
from mpl_toolkits.mplot3d import Axes3D

data = sio.loadmat(r'C:\Users\JAH\Desktop\组\random walk\classic2D\classic2D.mat')

p = np.array(data["p"])
Nstep = int(data["Nstep"])
site = np.array(data["site"])
X = np.array(data["X"])[0]
Y = np.array(data["Y"])[0]
X,Y = np.meshgrid(X,Y)


#> 观察粒子分布随时间(步数)的演化

fig, ax1 = plt.subplots(figsize=(16,9),subplot_kw={'projection':'3d'})

def update(i):

    ax1.clear()    #清除上一帧的axes对象

    g = ax1.plot_wireframe(X,Y,p[i],rstride=1,cstride=1,linewidth=0.2,color="b")

    ax1.set_title("2D classic random walk")
    ax1.set_xlabel('X')
    ax1.set_ylabel('Y')
    ax1.set_zlabel('p')

    ax1.set_zlim(0,1)

    return ax1


#> 单个粒子随机游走

# fig, ax1 = plt.subplots(figsize=(8,8))
# cMap = colors.ListedColormap(['white','black']) #自定义颜色 黑白色
# norm = mpl.colors.Normalize(0,1)

# def update(i):

#     ax1.clear()    #清除上一帧的axes对象

#     g = ax1.pcolor(X, Y, site[i], norm=norm, cmap=cMap)

#     ax1.set_title("2D classic random walk")
#     ax1.set_xlabel('X')
#     ax1.set_ylabel('Y')

#     ax1.set_xlim(-Nstep/10,Nstep/10)
#     ax1.set_ylim(-Nstep/10,Nstep/10)

#     return ax1


anim = FuncAnimation(fig, update, frames=np.arange(0, Nstep), interval=50, blit=False, repeat=False)
# plt.show()
anim.save(r'C:\\Users\\JAH\\Desktop\\组\\random walk\\classic2D\\2D classic probability distribution.gif')